lib/MatchingAlgorithm.cc

```diff
@@ -388,3 +389,102 @@ std::cout << stmp.str();
 #endif
 }
 
+Matching NProposingTiesMatching::compute_matching() {
+  FreeListType free_list; 
+  std::map<VertexPtr, VertexBookkeeping> bookkeep_data;
+  std::shared_ptr<BipartiteGraph> G = get_graph();
+  auto M = Matching(is_A_proposing());
+  
+  // choose the partitions from which the vertices will propose
+  const auto& proposing_partition = is_A_proposing() ? G->get_A_partition()
+                                                     : G->get_B_partition();
+
+  // set the level of every vertex in the proposing partition to 0
+  // mark all proposing vertices free (by pushing into the free_list)
+  // and vertices from the opposite partition implicitly free
+  for (auto &it : proposing_partition) {
+    auto v = it.second;
+    free_list.push(v);
+    int pref_list_size = v->get_preference_list().size();
+    int residual = v->get_upper_quota();
+    bookkeep_data[v] = VertexBookkeeping(0, pref_list_size, 0, residual);
+  }
+
+  // there is at least one vertex in the free list
+  while (not free_list.empty()) {
+    // arbitrary vertex in free list
+    auto u = remove_from_free_list(free_list, bookkeep_data);
+    const auto &u_pref_list = u->get_preference_list();
+    auto &u_data = bookkeep_data[u];
+    // if u^l can have a partner and hasn't exhausted its preference list
+    while (u_data.residual > 0 and not u_data.is_exhausted()) {
+      // Highest ranked vertex to whom u has not yet proposed.
+      VertexPtr v;
+      if (u_pref_list.is_tied(u_data.begin)) {
+        v = u_pref_list.get_ties(u_data.begin)[u_data.tied_index].vertex;
```

Comment:
Break this down into two parts, first part gets the ties and stores it into a variable. The second part gets the vertex at the tied index. It makes it much easier to read.